server/coordinator/node_picker.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package coordinator
+
+import (
+	"context"
+	"crypto/rand"
+	"math/big"
+
+	"github.com/CeresDB/ceresmeta/server/cluster"
+	"github.com/pkg/errors"
+)
+
+type NodePicker interface {
+	PickNode(ctx context.Context, clusterName string) (cluster.RegisteredNode, error)
+}
+
+type RandomNodePicker struct {
+	clusterManager cluster.Manager
+}
+
+func (p *RandomNodePicker) PickNode(ctx context.Context, clusterName string) (cluster.RegisteredNode, error) {
+	nodes, err := p.clusterManager.ListRegisterNodes(ctx, clusterName)
+	if err != nil {
+		return cluster.RegisteredNode{}, err
+	}
+
+	var onlineNodes []cluster.RegisteredNode
+	// Filter invalid nodes.
+	for _, node := range nodes {
+		if !node.IsOnline() {
+			onlineNodes = append(onlineNodes, node)
+		}
+	}
+
+	selectNodeIndex, err := rand.Int(rand.Reader, big.NewInt(int64(len(onlineNodes))))
+	if err != nil {
+		return cluster.RegisteredNode{}, errors.WithMessage(err, "generate random node index")
+	}
```

Comment:
Actually, we can avoid the slice creation:
```go
numOnlineNodes := 0
for idx := range nodes {
   if nodes[idx].IsOnline() {
       numOnlineNodes ++
   }
}
selectedIdx, err := rand.Int(rand.Reader, big.NewInt(numOnlineNodes))
for {
    if selectedIdx >= len(nodes) {
       // No valid node is found.
        return -1, nil
    }
    
    if nodes[selectedIdx].IsOnline() {
        return selectedIdx
    }
    selectedIdx++
}
```

Reply:
The above implementation will always return the same node in some cases, I tweaked a bit to fix this.